[2단계 - 로또(수동)] 쿤(홍동건) 미션 제출합니다.

Hongdonggeon opened the pull request:
안녕하세요 스티치!! 쿤입니다!!
리뷰를 너무 친절하게 달아주셔서 너무 많이 배울 수 있었던 것 같습니다!! 감사합니다 😭
static의 장단점, Enum의 사용법, Generator의 재사용, View의 싱글턴패턴화, 캐싱전략 그리고 사용범위에 따라 접근지정자를 잘 활용하는 것에 대해 많이 배울 수 있었던 것 같아요!!
평소에는 private이랑 public만 이용하였는데, 같은 패키지내에서만 사용하는 상수나 메서드 같은 경우에 default로 지정해 사용하는 것이 새로웠던 것 같아요!!

리뷰 반영과 2단계를 진행하면서 궁금한점들에 대해 정리해보았습니다!!

### 기존의 클래스형식 그대로 반환하고 있는데 dto로 구현한 이유가 있을까요?
기존의 클래스 형식 그대로 반환을 하고 있지만, VIEW와 Controller에서 독립적인 구조를 만들어주기 위함이였어요!! view가 현재처럼 하나의 application에 포함되어 있는 것이 아닌 view와 Controller에서 api로 통신을 하는 경우에, 통계 결과 출력에 필요한 데이터들을 한번의 통신으로 모두 보내주기 위헤 RankDto라는 클래스를 추가하였습니다.

### 기능 요구 사항을 보고, 어떤 도메인이 있을까 고민하는 과정속에서 어떤 기준으로 도메인을 나누는지 궁금합니다!
이런 고민을 하게 된게, Player 객체가 정말 필요할까라는 생각이 들었던 것 같아요. 로또 게임에 있어서 Player 객체가 여러번 생성될 일이 없기 때문에, 굳이 필요 없겠다는 생각이 드네요. (이 로또 프로그램이 진짜 출시가 된다면,,,? 이 프로그램을 사용하는 사람이 Player이지 않을까? 라는 생각이 들었습니다) 차라리 Money는 정말 Controller 메서드에서 관리가 되고, 차라리 List<Lotto>를 관리하는 Lottos라는 일급 컬렉션을 만들까도 고민이 정말 되었던 것 같아요.

Juhyung990122 committed:
docs: 기능 구현 목록 작성

Juhyung990122 committed:
feat: 입력 기능 구현

Juhyung990122 committed:
feat: 입력 기능 구현

Juhyung990122 committed:
feat: Money 객체 생성

Juhyung990122 committed:
feat: Money 객체 생성

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
feat: LottoNumber 객체 생성

Juhyung990122 committed:
feat: LottoNumber 객체 생성

Juhyung990122 committed:
feat: 로또 구매 기능 구현

생성자 안에 복잡한 메소드가 들어가도 될지 고민되었습니다.

Juhyung990122 committed:
feat: 로또 구매 기능 구현

생성자 안에 복잡한 메소드가 들어가도 될지 고민되었습니다.

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
feat: DTO 추가

Juhyung990122 committed:
feat: DTO 추가

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
feat: WinningLotto 객체 생성

Juhyung990122 committed:
feat: WinningLotto 객체 생성

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
feat: Winning Lotto 당첨번호와 보너스볼 간 중복검사

Juhyung990122 committed:
feat: Winning Lotto 당첨번호와 보너스볼 간 중복검사

Juhyung990122 committed:
refactor: 예외 메세지 별도 클래스로 분리

Juhyung990122 committed:
refactor: 예외 메세지 별도 클래스로 분리

Juhyung990122 committed:
feat: 로또 구매 및 당첨번호 세팅 기능 구현

Juhyung990122 committed:
feat: 로또 구매 및 당첨번호 세팅 기능 구현

Juhyung990122 committed:
feat: 로또 판정 로직 및 전체 판정 구현

Juhyung990122 committed:
feat: 로또 판정 로직 및 전체 판정 구현

Juhyung990122 committed:
feat: 보너스볼 판정 로직 및 Result 객체 구현

Juhyung990122 committed:
feat: 보너스볼 판정 로직 및 Result 객체 구현

Juhyung990122 committed:
feat: 2,3등은 보너스볼과 일치하는 숫자의 갯수를 기준으로, 나머지 등수는 일치하는 숫자의 갯수만으로 등수를 판정하는 로직 구현

Juhyung990122 committed:
feat: 2,3등은 보너스볼과 일치하는 숫자의 갯수를 기준으로, 나머지 등수는 일치하는 숫자의 갯수만으로 등수를 판정하는 로직 구현

Juhyung990122 committed:
feat: 결과 출력 기능 구현

Juhyung990122 committed:
feat: 결과 출력 기능 구현

Juhyung990122 committed:
feat: 수익률 계산 로직 및 출력 기능 구현

Juhyung990122 committed:
feat: 수익률 계산 로직 및 출력 기능 구현

Juhyung990122 committed:
refactor: 수익률 계산 로직 분리 및 위치 변경

Juhyung990122 committed:
refactor: 수익률 계산 로직 분리 및 위치 변경

Juhyung990122 committed:
refactor: 프린트 할 결과물 생성 로직 변경

Juhyung990122 committed:
refactor: 프린트 할 결과물 생성 로직 변경

Juhyung990122 committed:
refactor: 결과 출력 로직 분리

Juhyung990122 committed:
refactor: 결과 출력 로직 분리

Juhyung990122 committed:
refactor: 결과 리스트 생성 로직 이동

Juhyung990122 committed:
refactor: 결과 리스트 생성 로직 이동

Juhyung990122 committed:
refactor: 매직넘버 상수화

Juhyung990122 committed:
refactor: 매직넘버 상수화

Juhyung990122 committed:
refactor: 입력 유효성 검사 static으로 변경 및 적용

Juhyung990122 committed:
refactor: 입력 유효성 검사 static으로 변경 및 적용

Juhyung990122 committed:
style: 코드 포멧팅

Juhyung990122 committed:
style: 코드 포멧팅

Juhyung990122 committed:
style: 코드포멧팅

Juhyung990122 committed:
style: 코드포멧팅

Juhyung990122 committed:
refactor: 로또 생성 기능 인터페이스로 분리

Juhyung990122 committed:
refactor: 로또 생성 기능 인터페이스로 분리

Juhyung990122 committed:
style: 코드 포멧팅

Juhyung990122 committed:
style: 코드 포멧팅

Juhyung990122 committed:
refactor: 출력 개행 추가

Juhyung990122 committed:
refactor: 출력 개행 추가

Juhyung990122 committed:
refactor: 패키지 분리

Juhyung990122 committed:
refactor: 패키지 분리

Hongdonggeon committed:
fix: controller import 문 추가

Hongdonggeon committed:
fix: controller import 문 추가

Hongdonggeon committed:
refactor : 매직넘버 상수화

Hongdonggeon committed:
refactor : 매직넘버 상수화

Hongdonggeon committed:
refactor : 일급 컬렉션의 List 불변 객체로 만들기

생성자의 파라미터로 받는 List를 그대로 멤버 변수에 주입하지 않고, new ArrayList()로 새로운 주소값을 할당받음
Collections.unmodifiableList()를 이용해 밖에서 List의 값이 바뀌지 않게 막는다.

Hongdonggeon committed:
refactor : 일급 컬렉션의 List 불변 객체로 만들기

생성자의 파라미터로 받는 List를 그대로 멤버 변수에 주입하지 않고, new ArrayList()로 새로운 주소값을 할당받음
Collections.unmodifiableList()를 이용해 밖에서 List의 값이 바뀌지 않게 막는다.

Hongdonggeon committed:
refactor : Dto 생성자 private 으로 접근지정자 변경

Dto 생성자를 자기 자신의 클래스에서만 사용하기 때문에 private 접근지정자로 변경해줌
외부에서 Dto 객체 생성해주는것도 막아줄 수 있다.

Hongdonggeon committed:
refactor : Dto 생성자 private 으로 접근지정자 변경

Dto 생성자를 자기 자신의 클래스에서만 사용하기 때문에 private 접근지정자로 변경해줌
외부에서 Dto 객체 생성해주는것도 막아줄 수 있다.

Hongdonggeon committed:
refactor : 메서드 접근지정자 default로 변경

LottoDto, RankDto의 from 메서드가 같은 패키지내에서만 사용하기 때문에 접근 지정자를 default로 변경해줌

Hongdonggeon committed:
refactor : 메서드 접근지정자 default로 변경

LottoDto, RankDto의 from 메서드가 같은 패키지내에서만 사용하기 때문에 접근 지정자를 default로 변경해줌

Hongdonggeon committed:
refactor : 당첨 로또와 사용자 로또 비교하는 메서드와, 보너스볼과 사용자 로또 비교해주는 메서드 위치 변경

WinningLotto 객체를 가져와서 당첨번호와 보너스볼이란 값을 꺼내서 비교하는 것이 아닌, WinningLotto 객체에게 LottoNumber 를 포함하고 있는지 메시지를 보내주는 형식으로 변경

Hongdonggeon committed:
refactor : 당첨 로또와 사용자 로또 비교하는 메서드와, 보너스볼과 사용자 로또 비교해주는 메서드 위치 변경

WinningLotto 객체를 가져와서 당첨번호와 보너스볼이란 값을 꺼내서 비교하는 것이 아닌, WinningLotto 객체에게 LottoNumber 를 포함하고 있는지 메시지를 보내주는 형식으로 변경

Hongdonggeon committed:
refactor : List<String>을 List<Integer> 로 변환해주는 역할 View에 위임

View의 역할은 구분자를 분리하고, 공백을 제거해주는 등과 같은 데이터를 최소한의 정리된 형태로 만들어, Controller로 넘겨주는 것까지라고 생각된다.

Hongdonggeon committed:
refactor : List<String>을 List<Integer> 로 변환해주는 역할 View에 위임

View의 역할은 구분자를 분리하고, 공백을 제거해주는 등과 같은 데이터를 최소한의 정리된 형태로 만들어, Controller로 넘겨주는 것까지라고 생각된다.

Hongdonggeon committed:
feat : 반복되는 LottoNumber 인스턴스 캐싱하기

LottoNumber 1에서 45까지 클래스 필드 List<LottoNumber>에 추가하여, LottoNumber 객체가 필요할 때, valueOf() , values()로 꺼내서 사용한다.
모든 클래스는 같은 LottoNumber 인스턴스를 사용하므로, equals와 hashcode도 필요없게 되었다

Hongdonggeon committed:
feat : 반복되는 LottoNumber 인스턴스 캐싱하기

LottoNumber 1에서 45까지 클래스 필드 List<LottoNumber>에 추가하여, LottoNumber 객체가 필요할 때, valueOf() , values()로 꺼내서 사용한다.
모든 클래스는 같은 LottoNumber 인스턴스를 사용하므로, equals와 hashcode도 필요없게 되었다

Hongdonggeon committed:
remove : LottoNumberFactory 삭제

LottoNumber의 List<LottoNumber>를 모든 클래스에서 사용하게 하면서, 따로 1 ~45 의 숫자를 리스트에 넣어주는 작업이 필요 없어졌다

Hongdonggeon committed:
remove : LottoNumberFactory 삭제

LottoNumber의 List<LottoNumber>를 모든 클래스에서 사용하게 하면서, 따로 1 ~45 의 숫자를 리스트에 넣어주는 작업이 필요 없어졌다

Hongdonggeon committed:
refactor : Player 로또 구매 역할 분리

Player에 로또를 구매할 수 있는지여부와 로또를 구매하는 역할추가
로또 구매 기능에 있어서 Player는 전달받는 (생성된)로또를 로또 보관함에 추가하한다.

Controller에서 LottoGenerator를 이용하여 로또를 생성한다.

Hongdonggeon committed:
refactor : Player 로또 구매 역할 분리

Player에 로또를 구매할 수 있는지여부와 로또를 구매하는 역할추가
로또 구매 기능에 있어서 Player는 전달받는 (생성된)로또를 로또 보관함에 추가하한다.

Controller에서 LottoGenerator를 이용하여 로또를 생성한다.

Hongdonggeon committed:
feat : 구매한 모든 로또의 결과를 기록하는 클래스 추가

이전에는 Enum에서 hitcount라는 변수를 사용하여, 로또의 Rank 기록들을 관리하였는데 Enum의 필드는 모두 상수로 관리되어야 한다.
그래서 ResultStatus라는 당첨 통계를 기록하고 관리하는 클래스를 추가하였다.

Hongdonggeon committed:
feat : 구매한 모든 로또의 결과를 기록하는 클래스 추가

이전에는 Enum에서 hitcount라는 변수를 사용하여, 로또의 Rank 기록들을 관리하였는데 Enum의 필드는 모두 상수로 관리되어야 한다.
그래서 ResultStatus라는 당첨 통계를 기록하고 관리하는 클래스를 추가하였다.

Hongdonggeon committed:
refactor : InputView, OutputView 싱글턴 패턴으로 관리

view의 모든 메서드들이 static 메서드로 선언이 되어있었는데, static은 GC의 대상이 아니기 때문에 view의 인스턴스는 static으로 관리하고 메서드들은 인스턴스로 관리한다.

Hongdonggeon committed:
refactor : InputView, OutputView 싱글턴 패턴으로 관리

view의 모든 메서드들이 static 메서드로 선언이 되어있었는데, static은 GC의 대상이 아니기 때문에 view의 인스턴스는 static으로 관리하고 메서드들은 인스턴스로 관리한다.

Hongdonggeon committed:
refactor : 인터페이스에서 원하는 추상 메서드만 몸체를 구현하고자 어댑터 클래스 추가

인터페이스 추상 메서드를 오버로딩하여 선언하였고 인터페이스를 원하는 메서드만 구현하고자 어댑터 클래스를 추가하였다.
AutoLottoGenerator에서는 parameter가 없는 generateLotto만, ManualLottoGenerator에서는 parameter가 있는 generateLotto를 오버라이딩 하였따

Hongdonggeon committed:
refactor : 인터페이스에서 원하는 추상 메서드만 몸체를 구현하고자 어댑터 클래스 추가

인터페이스 추상 메서드를 오버로딩하여 선언하였고 인터페이스를 원하는 메서드만 구현하고자 어댑터 클래스를 추가하였다.
AutoLottoGenerator에서는 parameter가 없는 generateLotto만, ManualLottoGenerator에서는 parameter가 있는 generateLotto를 오버라이딩 하였따

Hongdonggeon committed:
refactor: 상수 접근지정자 변경

Hongdonggeon committed:
refactor: 상수 접근지정자 변경

Hongdonggeon committed:
refactor: 메서드 위치 변경

코드가 사용하는 곳에 가까이 붙어있으면 가독성이 더 좋아진다

Hongdonggeon committed:
refactor: 메서드 위치 변경

코드가 사용하는 곳에 가까이 붙어있으면 가독성이 더 좋아진다

Hongdonggeon committed:
refactor : 상수 접근지정자 변경

Hongdonggeon committed:
refactor : 상수 접근지정자 변경

Hongdonggeon committed:
refactor : 상수 네이밍 컨벤션 적용

Hongdonggeon committed:
refactor : 상수 네이밍 컨벤션 적용

Hongdonggeon committed:
refactor : 상수 접근지정자 변경

Money, Player 2곳에서 같은 상수를 관리하고 있어서 Money에서 접근지정자를 default로 변경하여 같은 패키지내에서 사용하게 해주었다.

Hongdonggeon committed:
refactor : 상수 접근지정자 변경

Money, Player 2곳에서 같은 상수를 관리하고 있어서 Money에서 접근지정자를 default로 변경하여 같은 패키지내에서 사용하게 해주었다.

Hongdonggeon committed:
refactor : InputValidator의 역할 InputView에게 위임

InputValidator이 하는일이 많지 않기 때문에 InputView가 InputValidator의 역할을 가지는 것이 더 효율적이고 응집도도 높아진다.

Hongdonggeon committed:
refactor : InputValidator의 역할 InputView에게 위임

InputValidator이 하는일이 많지 않기 때문에 InputView가 InputValidator의 역할을 가지는 것이 더 효율적이고 응집도도 높아진다.

Hongdonggeon committed:
refactor : LottoNumber 컬렉션 Map으로 변경

List 로 Cache를 만든 경우에, 리스트의 인덱스에 의존하게 된다.
Map으로 구현을 하게 되면, 인덱스에 의존하지 않게 된다.

Hongdonggeon committed:
refactor : LottoNumber 컬렉션 Map으로 변경

List 로 Cache를 만든 경우에, 리스트의 인덱스에 의존하게 된다.
Map으로 구현을 하게 되면, 인덱스에 의존하지 않게 된다.

Hongdonggeon committed:
feat: cache 에 로또 번호가 삭제되었을 경우에, cache에 다시 저장하게 함

Hongdonggeon committed:
feat: cache 에 로또 번호가 삭제되었을 경우에, cache에 다시 저장하게 함

Hongdonggeon committed:
feat : Lotto 클래스에 lottoNumber를 포함 여부를 보여주는 메서드 추가

Hongdonggeon committed:
feat : Lotto 클래스에 lottoNumber를 포함 여부를 보여주는 메서드 추가

Hongdonggeon committed:
feat : Rank에 2등인지만 확인하는 메서드 추가

Rank와 Result에 isSecondRank라는 메서드를 추가하여 judgeResult()에서 검증
Rank의 hitCount 인스턴스 변수도 삭제

Hongdonggeon committed:
feat : Rank에 2등인지만 확인하는 메서드 추가

Rank와 Result에 isSecondRank라는 메서드를 추가하여 judgeResult()에서 검증
Rank의 hitCount 인스턴스 변수도 삭제

Hongdonggeon committed:
refactor : Interface default 메서드 추가

이전에는 adapter 클래스를 추가하여 오버라이딩하고 싶은 메서드만 골라 오버라이딩하였다.
기존의 로또 자동 생성기능은 추상 메서드로 구현하고 새로 추가된 로또 수동생성 메서드는 default 메서드로 구현
AutoLottoGenerator와 ManualLottoGenerator에서 알맞지 않은 기능을 사용할시 예외 발생

Hongdonggeon committed:
refactor : Interface default 메서드 추가

이전에는 adapter 클래스를 추가하여 오버라이딩하고 싶은 메서드만 골라 오버라이딩하였다.
기존의 로또 자동 생성기능은 추상 메서드로 구현하고 새로 추가된 로또 수동생성 메서드는 default 메서드로 구현
AutoLottoGenerator와 ManualLottoGenerator에서 알맞지 않은 기능을 사용할시 예외 발생

Hongdonggeon committed:
feat : 로또 구매 기능 구매할 수 있는지 유효성 검증 기능 추가

while 문에서 로또를 구매할 수 있는지 검사해주고 있으나, while 문 밖에서 이용할 시 돈이 부족해도 구매가 되는 형태였다. 그래서 purchaseLotto에 돈의 상태에 대해 유효성검증을 해주는 기능을 추가했다.

Hongdonggeon committed:
feat : 로또 구매 기능 구매할 수 있는지 유효성 검증 기능 추가

while 문에서 로또를 구매할 수 있는지 검사해주고 있으나, while 문 밖에서 이용할 시 돈이 부족해도 구매가 되는 형태였다. 그래서 purchaseLotto에 돈의 상태에 대해 유효성검증을 해주는 기능을 추가했다.

Hongdonggeon committed:
feat : Money 1000원 단위 유효성 검증 추가

Hongdonggeon committed:
feat : Money 1000원 단위 유효성 검증 추가

Hongdonggeon committed:
feat : 수동 로또 뷰 구현

Hongdonggeon committed:
feat : 수동 로또 뷰 구현

Hongdonggeon committed:
feat : 수동로또 갯수, 자동로또 갯수를 관리하는 클래스 추가

수동로또 음수를 입력할 경우 예외 발생
수동로또가 구입 금액으로 구매할 수 있는 로또수를 초과하였을 때, 자동로또 갯수가 음수가 되므로 예외발생

Hongdonggeon committed:
feat : 수동로또 갯수, 자동로또 갯수를 관리하는 클래스 추가

수동로또 음수를 입력할 경우 예외 발생
수동로또가 구입 금액으로 구매할 수 있는 로또수를 초과하였을 때, 자동로또 갯수가 음수가 되므로 예외발생

Hongdonggeon committed:
feat : 구입 금액으로 구매할 수 있는 로또 총 갯수 반환하기

Hongdonggeon committed:
feat : 구입 금액으로 구매할 수 있는 로또 총 갯수 반환하기

Hongdonggeon committed:
feat : 수동 로또 갯수에 대한 예외 메세지 추가

Hongdonggeon committed:
feat : 수동 로또 갯수에 대한 예외 메세지 추가

Hongdonggeon committed:
feat : 수동 로또 구매 기능 추가

Player.java
   selectLottoCount() : 수동 로또 구매갯수를 파라미터로 받아 LottoCount객체를 생성하면서 유효성 검증을 한다.

LottoController.java
   purchaseManualLotto() : InputView에서 입력받은 숫자를 ManualLottoGenerator로 로또를 생성한다.

Hongdonggeon committed:
feat : 수동 로또 구매 기능 추가

Player.java
   selectLottoCount() : 수동 로또 구매갯수를 파라미터로 받아 LottoCount객체를 생성하면서 유효성 검증을 한다.

LottoController.java
   purchaseManualLotto() : InputView에서 입력받은 숫자를 ManualLottoGenerator로 로또를 생성한다.

Hongdonggeon committed:
refactor : view 수정

InputView, OutputView 양식에 맞게 개행문자 추가

Hongdonggeon committed:
refactor : view 수정

InputView, OutputView 양식에 맞게 개행문자 추가

Hongdonggeon committed:
refactor : 수동로또 생성할 때 정렬시키기

Hongdonggeon committed:
refactor : 수동로또 생성할 때 정렬시키기

Hongdonggeon committed:
refactor : purchaseLotto의 파라미터로 외부에서 LottoGenerator 구현체를 생성해서 주입해주면서 메서드를 재사용하는 형식으로 리팩토링

Hongdonggeon committed:
refactor : purchaseLotto의 파라미터로 외부에서 LottoGenerator 구현체를 생성해서 주입해주면서 메서드를 재사용하는 형식으로 리팩토링

Hongdonggeon committed:
test : 수동 로또 갯수에 대한 테스트

Hongdonggeon committed:
test : 수동 로또 갯수에 대한 테스트

Hongdonggeon committed:
test : money 로직에 대한 테스트

Hongdonggeon committed:
test : money 로직에 대한 테스트

Hongdonggeon committed:
test : LottoGenerator 구현체 예외 테스트

Hongdonggeon committed:
test : LottoGenerator 구현체 예외 테스트

Hongdonggeon committed:
docs : Rank 기능에 대한 테스트

Hongdonggeon committed:
docs : Rank 기능에 대한 테스트

Hongdonggeon committed:
test : ResultStatus 기능에 대한 테스트

Hongdonggeon committed:
test : ResultStatus 기능에 대한 테스트

Hongdonggeon committed:
conflict 헤결

Hongdonggeon committed:
conflict 해결

Hongdonggeon commented on src/main/java/controller/LottoController.java:
```diff
@@ -19,9 +19,23 @@
+        return player.selectLottoCount(ManualLottoCount);
+    }
+
+    public LottosDto purchaseLotto(LottoGenerator lottoGenerator, List<List<Integer>> manualNumber) {
```
purchaseLotto를 자동생성일때, 수동 생성일 때 둘다 재사용하기 위해 메서드를 이렇게 바꾸어보았는데요!! AutoLotoGenerator의 generate()와 ManualLottoGenerator에서 사용되는 generate()의 시그니처가 달라서 재사용되는 구조 만들기가 쉽지 않았던 것 같아요,, 코드에서 null을 사용하는 것에 대해 걱정도 되고요.. 이렇게 if문을 분기로 나누는 것보다 메서드를 나누는게 더 좋은 방법이지 않을까라는 생각도 들고요,, 어떻게 하는게 좋을까요?

lxxjn0 replied:
시그니처를 어떻게 고민을 해봐도 같게 만들기 어려울 경우는 '정말 두 클래스가 같은 인터페이스로 관리가 되어야 하는가?'에 대해서 의문을 가져보면 좋을 것 같아요!! 

물론 생성을 한다는 개념은 같지만 하나는 추가적인 매개변수가 없이 생성이 가능하고 하나는 추가적인 매개변수를 사용해서 생성을 해야 합니다. 그렇다면 정말 두 클래스가 동일한 generate를 하는 것일까요?? 이런 고민들을 해보면서 정말 동일한 인터페이스로 묶어야 하는 것인가? 를 생각해볼 수 있을 것 같아요!!

추가로 만약 위와 같은 고민을 한 이후에도 '이 둘은 동일한 generate가 맞다'라는 생각이 든다면 그때는 어떻게 동일한 인터페이스화를 시킬 수 있는지에 대해서 고민을 해야할 것 같아요!! 예를 들어 현재는 AutoLottoGenerator와 ManualLottoGenerator의 가장 큰 문제점은 숫자를 생성하는 부분입니다. 그렇다면 숫자를 생성하는 부분을 다른 클래스(+ 인터페이스)로 추출할 수는 없을까?? 등을 고민해볼 수도 있을 것 같아요!! 

한번 이런 고민들을 해보면서 코드를 개선해본다면 좋은 답을 찾을 수 있을 것 같습니다!!!

lxxjn0 replied:
추가로 현재의 코드를 유지하는 것은 인터페이스 추상화를 하지 않은 코드보다 더 안 좋은 코드라고 생각합니다!! 적절한 상황에서 추상화를 하는 것은 좋지만 지금은 추상화를 하기 위해서 코드를 작성한다는 느낌이 강해요. 

추상화는 코드를 유지보수하기 좋게 관리하기 위한 방법입니다. 말 그대로 방법이에요. 코드를 '객체지향적으로 짜기 위한 목적'을 달성하기 위한 수단으로 추상화가 사용되어야 합니다. '추상화를 사용하려는 목적'을 위해서 코드를 희생해서는 안됩니다!!

이번 기회에 쿤이 해당 부분을 한번 더 인지하면 앞으로 많은 도움이 될 것 같아요!! MVC 패턴을 사용해서 멋있게 레이어화 시키는 것, DTO를 사용해서 깔끔하게 view와 도메인을 분리해보는 것, 인터페이스 추상화를 통해 클래스를 깔끔하게 관리하는 것, 이런 것들 모두 절대 나쁜 것은 아닙니다!! 하지만 주객전도가 되면 안돼요.

MVC 패턴도, DTO 사용도, 인터페이스 추상화도 모두 '코드를 유지보수하기 좋게 관리하려는 목적'을 달성하기 위한 '수단'입니다. 만약 해당 수단을 사용하였을 때 오히려 코드가 유지보수하기 적절하지 않다면 이는 잘못 사용한 것이고 사용하지 말아야 했던 것입니다.

지금 쿤의 모든 코드가 그렇다는 것은 절대 아니에요!! 혹시나 오해하실까 싶어서 말씀드립니다 ㅠㅠ. 단지 앞으로 이런 부분들에 마주했을 때 쿤이 생각해야 하는 부분은 '내가 코드를 멋지게, 보기 깔끔하게 짜기 위해서 무엇을 희생하였는가. 과연 내가 희생한 부분이 단순히 코드를 깔끔하게 작성하고 싶은 나의 욕심 때문인가 아니면 정말 좋은 코드, 유지보수하기 좋은 구조를 가진 코드를 만들려는 의도 때문이가' 입니다. 쿤이 이러한 생각을 할 수 있는 계기가 되었으면 해서 조금 더 다양한 예시를 들어서 말씀드렸어요!! :)

이번 기회에 한번 이런 부분들에 대한 스스로의 관점을 가져보는 기회가 된다면 정말 좋을 것 같습니다!! 💯

lxxjn0 commented on src/main/java/controller/LottoController.java:
```diff
@@ -19,9 +19,23 @@
     private Player player;
     private WinningLotto winningLotto;
```
현재 컨트롤러에서 player와 winningLotto는 생성 시점에서 초기화 되지 않고 있습니다. 그러면 사용자가 해당 필드를 초기화하지 않은 상태에서 다른 메서드를 사용할 수 있는 상황을 제공하게 되고, 그런 상황에서 NPE와 같은 에러가 발생할 수 있습니다. 이는 불변 필드를 가지지 않음으로 발생할 수 있는 문제점입니다.

이처럼 클래스를 객체로 만들어서 사용하려고 한다면 최대한 불변으로 사용해야지 안전하게 사용할 수 있습니다. 만약 필드가 변경이 필요한 '어쩔 수 없는 상황'에 대해서는 가변 변수로 사용할 수 있습니다. 

즉 해당 필드들이 객체 입장에서 변경될 필요가 없는 값들이라면 생성 시점에서 불변하게 관리되어야 하며, 만약 그렇지 않은 값이라면 필드 변수로 관리되어야 할 값이 아닐 가능성이 큽니다!! 그런 관점에서 과연 해당 필드들이 정말 메서드로 초기화 되어야 하는 값들인가? 그렇지 않고 추가로 필드가 아니게 사용할 수도 있는가? 등을 이번에 한 번 추가적으로 고민해보면 좋을 것 같아요!!

Hongdonggeon replied:
이전에 Controller 메서드들을 작성을 할 때, Controller 메서드의 파라미터는 요청이 들어오는 값, 반환형은 View로 데이터를 보낼 값이라고 틀을 정해놓고 작성했던 것 같아요!! player와 winningLotto를 필드변수로 지정한 이유는 여러 메서드에서 사용하기 때문에 필드 변수로 선언을 했었던 것이구요!! 이번에는 리팩토링을 하면서 다시 봐보니 LottoController 객체가 생성이 될 때 필요한 값들은 없는 것 같아 필드 변수가 없게 설계를 해보았어요!! 여러 메서드에서 사용되는 객체는 메서드의 파라미터로 전달해 사용하게 해보았습니다

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumber.java:
```diff
@@ -3,18 +3,20 @@
     private static final int MINIMUM_LOTTO_NUMBER = 1;
     private static final int MAXIMUM_LOTTO_NUMBER = 45;
-    private static final List<LottoNumber> CACHE = new ArrayList<>();
+    private static final Map<Integer, LottoNumber> CACHE = new HashMap<>();
```
Map을 사용한 부분 👍🏼 이를 통해서 기존에 List로서 가질 수 있는 문제점이 충분히 해결된 것 같아요!!!

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumber.java:
```diff
@@ -3,18 +3,20 @@
+        for (int number = MINIMUM_LOTTO_NUMBER; number <= MAXIMUM_LOTTO_NUMBER; number++) {
+            CACHE.put(number, new LottoNumber(number));
         }
     }
```
쿤의 코드 중 valueOf 메서드에서 해당 Map에 존재하지 않을 경우 다시 생성자를 호출하고 있는데, 그 이유에 대해서 제가 곰곰히 고민을 해봤습니다. 그래서 '만약 CACHE에서 값이 지워지는 경우가 생기면 어떡하지? 그럴 경우는 범위에 맞는 LottoNumber일 경우는 다시 생성해서 추가해주자'와 같은 생각을 거쳐서 이런 valueOf 로직이 생긴게 아닐까 싶습니다.

그렇다면 만약 CACHE가 처음부터 Immutable하게 생성해서 관리된다면 어떨까요? 그렇다면 valueOf에서 해당 검증은 필요하지 않을 것입니다. 코드 상에서 불가능한 케이스기 때문이죠. 그리고 CACHE도 불변하게 관리되므로 값의 추가, 삭제에 대한 걱정을 할 필요가 없습니다!!

static 변수를 final로 두더라도 static 스코프 내에서만 초기화를 한다면 굳이 선언과 동시에 초기화를 할 필요가 없습니다!! 아래 코드와 같이 말이죠

```java
private static final List<String> CACHE;

static {
    CACHE = new ArrrayList();
}
```

이런 부분을 고려해서 한번 더 코드를 개선해보면 좋을 것 같아요!! 👍🏼

Hongdonggeon replied:
너무너무 좋은 방법인 것 같습니다 감사합니다!!

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumber.java:
```diff
@@ -25,23 +27,24 @@ public LottoNumber(int number) {
+        }
+    }
+
     public static LottoNumber valueOf(int number) {
```
현재는 valueOf 정적 팩토리 메서드 말고 생성자도 public으로 열려있습니다!! 그렇다면 사용자는 cache 된 값에서만 LottoNumber를 가져다 쓰는 것이 아니라 직접 생성해서 쓸 수 있어요!! 이런 부작용을 막으려면 어떻게 하면 될까요??

Hongdonggeon replied:
제가 생각한 의도로는요! 이전의 코드가 아래의 코드였어요!!
```java
    public static LottoNumber valueOf(int number) {
        LottoNumber lottoNumber = CACHE.get(number - 1);

        if (lottoNumber == null) {
            lottoNumber = new LottoNumber(number);
        }
        return lottoNumber;
    }
```
만약에 Cache에서 번호가 1인 LottoNumber가 삭제되었을 경우, 로또 번호 1을 고를 때, 숫자가 1인 LottoNumber 는 모두 다른 객체가 될 것입니다. 그래서 Cache 에 삭제 되어버린 번호 1인 LottoNumber를 새로 추가를 해주고 Cache에서 번호 1인 LottoNumber 를 반환되게 해주었어요!! 
어차피 1에서 45에서 벗어난 숫자는 LottoNumber 객체가 생성이 될 때, 예외를 발생하여 LottoNumber가 생성되지 못하게 막아 주지 않을까라고 생각했습니다

Hongdonggeon replied:
위의 Review를 보고 감동했습니다🥲  제 의도를 고민해주셔서 너무 감사드립니다!! Cache를 불변으로 만드는 것이 더 좋은 방법인 것 같아요!! 감사합니다

lxxjn0 commented on src/main/java/domain/Lotto/Lotto.java:
```diff
@@ -9,7 +9,7 @@
 
 public class Lotto {
```
현재 해당 클래스에서 judge라는 메서드가 사용되고 있는데 메서드 명이 조금 추상적이란 생각이 들었어요!! 

사용하는 사용자 입장에서 `Result result = lotto.judge(winningLotto)` 와 같이 사용되고 있는데 만약 로또와 당첨 로또 번호를 비교하는 행위를 하고 있다면 compare라는 메서드 명이 조금 더 자연스럽다고 생각해요. 실제 로직도 당첨 로또와 비교를 하는 로직이니깐 말이죠. 그래서 그에 따른 비교 결과가 반환되는 것도 자연스러울 것 같구요!!

해당 코멘트는 저의 주관적인 생각이라서 쿤이 한번 생각해보면 어떨까? 싶어서 남겨봤습니다!! ㅎㅅㅎ 한번 참고만 해주세요 ㅎㅎ

Hongdonggeon replied:
스티치의 의견에 동의합니다!! 다시 봐보니까 Result라는 클래스가 굳이 필요 없겠다라는 생각도 들었던 것 같아요!! Rank의 결과를 판단해주는 메서드 시그니처를 변경하여 Lotto의 compare메서드에서 바료 사용가능하게 변경해주었습니다. 그리고 Result의 클래스를 삭제했습니다!!

lxxjn0 commented on src/main/java/domain/Lotto/Lotto.java:
```diff
@@ -33,7 +33,12 @@ public Result judge(WinningLotto winningLotto) {
 
+    public boolean isContainLottoNumber(LottoNumber lottoNumber) {
+        return lotto.contains(lottoNumber);
+    }
```
해당 변수명도 `hasLottoNumber`를 생각해봤는데 어떤가요?? 조금 더 자연스러운 이름이라는 생각이 들어서 조심스래 남겨봅니다 :)

Hongdonggeon replied:
좋은 것 같아요!! has를 앞에 붙이는 것이 데이터를 가지고 있는지 확인하는 메서드 명에 있으면 좋을 것 같아요!!

lxxjn0 commented on src/main/java/domain/LottoGenerator/AutoLottoGenerator.java:
```diff
@@ -25,4 +26,10 @@ public Lotto generateLotto() {
+    @Override
+    public Lotto generateLotto(List<Integer> numbers) {
+        throw new IllegalArgumentException(ExceptionMessage.MANUAL_LOTTO_GENERATE_LIMIT);
+    }
```
이렇게 해당 클래스에서 사용될 수 없는 메서드라는 것을 말하고 싶을 땐 UnsupportedOperationException을 주로 사용합니다. 아래는 해당 예시로 java.util.immutableCollection의 코드입니다

<img width="983" alt="image" src="https://user-images.githubusercontent.com/48052622/156893620-30881728-8d16-4c54-8466-27bd50a15c00.png">

Hongdonggeon replied:
이 Exception, 일급 컬렉션에서 unmodifiable로 감쌌을 때 컬렉션에 데이터 추가할 때 발생했던 Exception이네요!! 감사합니다 ☺️

lxxjn0 commented on src/main/java/domain/player/Money.java:
```diff
@@ -21,10 +34,8 @@ public boolean isBiggerThanLottoPrice() {
-        if (amount < MINIMUM_PURCHASE_AMOUNT) {
-            throw new IllegalArgumentException(ExceptionMessage.NOT_ENOUGH_MONEY);
-        }
+    public int calculateTotalLottoCount() {
```
해당 메서드의 행위는 '구매 가능한 로또의 개수'와 동일한 것으로 보여져요. calculateTotalLottoCount 메서드 명은 그런 의도를 표현하기에는 '전체 로또 개수를 계산하다' 정도의 의미라서 조금은 설명이 부족한 것 같다고 느껴졌습니다!!
예로 `calculatePurchasableLottoCount` 와 같은 메서드 명이 좀 더 행위를 잘 설명하는 것 같아요!! 쿤이 한번 고민해보면 좋을 것 같습니다!! :)

Hongdonggeon replied:
동의합니다 메서드명 짓기 너무 힘드네요,,ㅠ 리뷰 반영했습니다 😀

lxxjn0 commented on src/main/java/domain/player/Player.java:
```diff
@@ -25,12 +25,20 @@ public boolean canBuyLotto() {
+    public LottoCountDto selectLottoCount(int manualLottoCount) {
+        int totalLottoCount = money.calculateTotalLottoCount();
+        LottoCount lottoCount = new LottoCount(manualLottoCount, totalLottoCount - manualLottoCount);
+        return LottoCountDto.of(lottoCount);
```
현재 코드는 DTO를 만들기 위해서 LottoCount 객체가 사용되고 있어요!! 과연 이러한 사용이 적절할까요?? 이러한 사용 용도로의 LottoCount라면 그냥 LottoCountDto에서 검증을 하면 되지 않을까요?? 

그렇다면 이러한 검증 로직을 가지는 LottoCount 객체만을 유지하는게 적절할까요 아니면 동일하게 검증 로직을 가지는 LottoCountDto 객체만을 유지하는게 적절할까요?? 쿤이 어떠한 선택이 적절한 것으로 보여지나요?

Hongdonggeon replied:
DTO는 정말 데이터를 보내기 위한 객체라고 생각을 해서 검증 로직이나 비즈니스 로직이 존재하면 안된다고 생각을 합니다. 그래서 LottoCountDto를 삭제하고 LottoCount객체만을 유지하는 것이 더 좋은 방법이라고 생각을 합니다

lxxjn0 commented on src/main/java/domain/player/Player.java:
```diff
@@ -25,12 +25,20 @@ public boolean canBuyLotto() {
+        }
         money.deductMoney();
         lottos.add(lotto);
     }
```
이렇게 메서드가 실행될 수 없는 상태일 경우는 IllegalStateException이 좀 더 적절할 것 같아요!!

그리고 추가적으로 현재 해당 메서드가 사용되는 부분을 보면 먼저 구매할 로또를 생성해서 메서드로 전달해주고 내부에서 금액을 차감합니다. 하지만 실제 프로세스를 생각해 본다면 구매 가능한 금액일 경우 금액을 차감하고 그 몫으로 로또를 player에게 전달해주는 것이 더 적절하지 않을까요?

Hongdonggeon replied:
동의합니다 돈이 차감되고 Lotto가 넘어와야 된다고 생각이 드네요!!

lxxjn0 commented on src/main/java/domain/player/Player.java:
```diff
@@ -40,7 +48,7 @@ public void purchaseLotto(Lotto lotto) {
 
     public double calculateIncomeRate(double totalIncome) {
-        return totalIncome / (lottos.size() * MINIMUM_PURCHASE_AMOUNT);
+        return totalIncome / (lottos.size() * Money.MINIMUM_PURCHASE_AMOUNT);
```
해당 라인을 본다면 MINIMUM_PURCHASE_AMOUNT라는 상수명은 적절하지 않은 것으로 보여요. 해당 로직을 풀어보면 전체 수익에서 (로또 개수 * 최소 구매 금액)을 나눕니다. 

상수의 역할은 읽는 사람으로 하여금 해당 값이 어떠한 의미를 가지는 지를 나타내기 위한 목적으로 사용하는 건데 현재 코드에서는 '로또 개수 * 로또 금액을 해야지 왜 최소 구매 금액을 곱하는거지?'라는 생각이 들 것 같아요!! 저 또한 그랬구요. 그런 관점에서 해당 상수 명을 다시 한번 고민해서 적절하게 바꿔보면 좋을 것 같습니다!!

Hongdonggeon replied:
맞는 것 같아요,, 최소 구매 금액이라는 표현이 약간 덜 와닿는 느낌도 있습니다. 로또 가격이라는 표현이 더 자연스러울 것 같네요!!

lxxjn0 commented on src/main/java/view/InputView.java:
```diff
@@ -1,44 +1,62 @@
     public int inputPurchaseAmount() {
         System.out.println(PURCHASE_AMOUNT_INPUT_MESSAGE);
         return Integer.parseInt(input());
     }
```
현재는 InputView에서 모든 변환 로직들이 동작하고 있어요. 그래서 LottoNumber를 가져오던, Money를 만들던, 모두 이미 변환된 정수를 사용해서 생성을 합니다. 그렇다면 이번에는 이러한 로직들을 도메인 객체가 가지도록 해보면 어떨까요??

간단한 예로 Java에서 LocalDateTime 객체를 생성할 때 문자열이나 정수 등 다양한 타입을 넘겨서 생성할 수 있습니다. 문자열을 넘겨주는 방법만 예시를 들어보면, 저희가 LocalDateTime을 문자열을 통해 생성할 때 전달하는 문자열이 적절한 시간 형식으로 이뤄졌는지, 유효한 시간 범위를 만족하는지 등을 모두 검사하고 생성을 할까요?? 저희는 생성할 때 어떤 문자열이든 LocalDateTime이 판단하도록 넘겨줍니다. 만약 해당 문자열이 LocalDateTime을 생성하기에 적절하지 않은 형태라면 알아서 적절한 예외를 던져주죠.

이는 LocalDateTime 클래스에서 풍부하게 생성 관련 메서드를 제공함으로서 우리가 얻는 이득입니다. 이처럼 쿤이 관리하는 많은 도메인들도 이렇게 사용자에게 풍부한 생성 메서드를 제공해주도록 개선해보면 어떨까요???

Hongdonggeon replied:
LocalDateTime 클래스를 찾아보았는데 of 메서드가 정말 많이 선언되어있더라구요!! 원래 제가 생각하는 View의 역할은 최소한의 데이터를 정리를 해서 (정수로 변경, 파싱) 보내주는 것이었는데 새로운 시도여서 흥미로웠던 것 같아요!! 어쩌면 지금의 도메인 형태가 View에 덜 의존적인 형태인 것 같기도 합니다!!

lxxjn0 reviewed (changes requested):
쿤, 2단계 미션 코드 잘 봤습니다!!

다양한 부분에서 어떻게 하면 더 좋은 코드를 작성할 수 있는지 고민한 부분들이 잘 보였던 것 같아요!! 이번 리뷰에서도 도메인 부분에 대한 고민 거리들에 대해서 주로 피드백을 남겨두었습니다!!

추가로 이번 리뷰 반영 때 dto를 다 없애도록 리팩토링 해보는 것은 어떨까? 조심스레 제안해봅니다. 

우테코 레벨 1 미션들의 목표는 어떻게 하면 객체지향적으로 코드를 작성해서 유지보수하기 좋은 코드를 만들 수 있을까를 고민하는 단계라고 생각합니다. 물론 이런 과정에 dto를 사용하는 것, 레이어드 아키텍처를 적절히 개선해보는 점 등은 무의미한 경험은 아니라고 생각해요. 

하지만 레벨 1 미션에서 정말 dto를 사용해야만 해결할 수 있는 미션들이 있는가?를 생각해보면 아니라고 생각합니다. 그리고 domain에서 view로 의존하는 방향은 절대 존재하면 안되지만 view에서 domain으로 향하는 의존성은 크게 상관이 없다고 생각해요. 그리고 지금 단계에서 dto를 어떤 단위로 작성할지, 어떻게 레이어를 더 잘 나눌지 등을 고민하는 것 보다는 도메인이 적절한지, 어떻게 하면 더 객체지향적인 코드를 짤 수 있는지에 대해서 고민하는게 더 맞다고 생각해요. 그리고 이런 과정이 쿤에게 오히려 더 많은 도움이 될 것입니다.

현재는 dto를 유지하고 사용하기 위해서 도메인 객체가 관리되는, 그런 상황까지 발생하는 것 같아요. 이는 이번 미션에서 관심을 기울여야 하는 부분에서 오히려 반대로 가는 모습이 아닌가 하는 생각이 들었습니다. 

그런 관점에서 쿤이 이번에 이러한 점들을 최대한 배제하고 정말 관심을 기울여야 하는 부분에 집중하면 좋겠다는 생각이 들어서 이러한 제안을 드려볼게요!! 쿤이 한번 곰곰히 생각해보고 한번 리뷰를 반영해보면 좋을 것 같습니다!!

Hongdonggeon committed:
refactor : LottoNumber Cache 불변 객체 만들기

이전에 1~45의 숫자가 삭제되었을 경우를 대비해 꺼낸 숫자가 null 일경우 새로 Map 에 저장을 했는데
애초에 Map에서 삭제가 안되게 UnmodifiableMap을 사용해 삭제를 막았다.

Hongdonggeon committed:
refactor : 메서드명 변경

LottoNumber(데이터)를 가지고 있는지의 여부를 나타내는 메서드명은 has로 시작하는것이 좋다.

Hongdonggeon committed:
refactor : 상수명 변경

Lotto 한장의 가격의 표현을 더 자연스러운 LOTTO_PRICE로 변경

Hongdonggeon committed:
refactor : Lotto와 WinningLotto 비교하는 메서드명 , 메서드 시그니처 변경

Result 클래스를 삭제하면서 Rank를 판단하는 파라미터를 hitCount, isHitBonusBall로 변경함으로서 Lotto에서 Rank 형으로 메서드 시그니처 변경

Hongdonggeon committed:
refactor : 메서드 명 변경

Hongdonggeon committed:
feat : 수동 Lotto와 자동 Lotto를 생성해주는 팩토리 메서드 추가

Hongdonggeon committed:
feat : 총 구매 로또를 관리해주는 Lottos 일급 컬렉션 추가

Hongdonggeon committed:
remove : Dto 삭제, LottoGenerator 인터페이스 삭제, Player 삭제

Dto 삭제 -> Controller에서 View에게 보내는 데이터를 Dto 로 보내지 않는다.

LottoGenerator -> LottoGenerator의 기능을 LottoFactory가 대체

Player 삭제 -> 돈과 전체로또를 관리하는 기능을 분배, 돈은 Controller에서만 관리하고 전체 로또는 Lottos가 관리

Hongdonggeon committed:
remove : Dto 삭제

Hongdonggeon committed:
remove : 수동로또 생성 메서드 접근지정자 변경

Controller에서 당첨 로또 생성할 때 사용하기 위해

Hongdonggeon committed:
rename : Money클래스 패키지 변경

Hongdonggeon committed:
test : 당첨 로또 생성할 때 사용하는 메서드 변경

Hongdonggeon committed:
test : judgeResult 메서드 시그니처가 변경하면서 불필요한 코드 삭제

Hongdonggeon committed:
test : LottoFactory의 기능에 대한 테스트

Hongdonggeon committed:
test : 당첨 로또 생성할 때, LottoFactory의 메서드로 변경

Hongdonggeon committed:
refactor : Money에서 필요없는 메서드 삭제

Hongdonggeon committed:
refactor : OutputView 메서드 시그니처 변경

Hongdonggeon committed:
feat : 수동 로또 생성하는 메서드 오버로딩 적용

파라미터가 String 일때도 내부에서 파싱해서 로또를 생성해주는 메서드

Hongdonggeon committed:
feat : Money 도메인에 String 파라미터로 Money 객체 생성해주는 메서드 추가

Hongdonggeon committed:
feat : LottoNumber valueOf 메서드 오버로딩

Hongdonggeon committed:
feat : LottoCount String 파라미터로 LottoCount 객체 생성해주는 메서드 추가

Hongdonggeon committed:
style : 코드 포맷팅

Hongdonggeon commented:
안녕하세요 스티치! 마지막 리뷰네요😭 
스티치의 리뷰를 보고 정말 많이 고민을 해봤어요
스티치의 말씀 처럼, 제가 적용해보고 싶은 인터페이스 추상화, Dto 사용 등 이러한 것들을 사용하기 위해 코드를 작성한 느낌이 들더라구요!!
그래서 이번에는 리팩토링 하면서, 각 클래스의 책임과 과연 이 클래스들은 필요한 클래스인가 고민하면서 리팩토링 해보았던 것 같아요!! 
그래서 사용자와 로또 프로그램의 요청, 응답을 중심으로 처음부터 Lotto 프로그램에 대해 생각을 해봤어요!! 

### 사용자와 로또 프로그램의 요청 응답
1. 사용자가 구입금액을 입력한다. -> 로또 프로그램은 돈의 유효성 검사를 해주고 돈을 충전해준다.
2. 사용자는 수동 로또 수를 입력한다. -> 수동 로또 수에 대한 유효성 검사를 해주고 수동 로또 갯수, 자동 로또 갯수를 기억하고 있는다.
3. 수동 로또 번호를 입력한다. -> 입력한 숫자대로 수동 로또를 생성해준 후 , 자동로또 갯수만큼 자동 로또를 생성해준다.
4. 수동 로또와 자동로또의 숫자들을 사용자에게 보내준다.
5. 당첨 번호과 보너스볼을 입력한다. -> 당첨 번호에 대한 유효성 검사를 해주고 당첨 번호를 만들어준다.
6. 사용자가 구매한 로또들과 당첨 번호들에 대해 비교해 결과를 산정한다. -> 당첨 통계 결과를 사용자에게 보여준다. 
7. 사용자가 충전한 돈과 당첨 금액으로 수익률을 계산한다. -> 당첨 통계 수익률을 사용자에게 보여준다. 

### 변화된 점
1. Result 클래스 삭제 
이전에는 사용자가 구매한 총 로또들을 일치 갯수와 보너스볼 일치여부를 필드로 가지고 있는 Result로 변환을 해서 Rank와 비교를 했었는데 Rank에서 일치 갯수, 보너스볼 일치 여부를 바로 파라미터로 받으면서 Result 클래스를 삭제해주게 되었습니다. 굳이 필요없는 작업이라고 생각을 하게 되었어요

2. Player -> Lottos 
사용자가 구매한 모든 로또와 돈을 관리해주던 Player객체였는데, Money는 Controller에서만 관리되게 하였고 모든 로또를 관리해주는 클래스로 Lottos를 추가하였습니다. 로또 프로그램에서는 사용자가 입력한 금액만큼 로또를 생성해주고 당첨 번호랑 비교만 해줄뿐 돈을 차감하는 등의 로직이 굳이 없어도 되겠다고 생각을 했습니다. 그리고 `purchaseLotto()` 의 로직이 조금 불안정하다고 생각을 했습니다.

3. LottoGenerator (인터페이스) -> LottoFactory
인터페이스의 메서드 시그니처가 달라서 추상화하기도 힘들었고, ManualLottoGenerator에서 자동 로또 생성 기능을 막아주는 등의 오히려 예외상황을 더 만든 것 같다는 생각을 했습니다. 좋은 경험이었다고 생각을 하지만, 예외 상황이 없는 LottoFactory로 만든 메서드가 더 안정적이고, LocalDateTime 처럼 다양한 파라미터가 들어와도 Lotto를 생성해주는 등의 구조가 더 좋다고 생각했습니다.

4. Dto 클래스 삭제
뷰와 도메인을 분리하는 것에 대한 집착이 있었던 것 같습니다. 그러다 보니 굳이 Dto를 만들지 않아도 되는 경우인데도 습관적으로 View에 데이터를 넘겨줄 때는 Dto를 만들었던 것 같아요! 리팩토링해보고 다시 봐보니 Dto로 사용하여 View에 데이터를 보냈던 것이 오히려 더 비효율적인 같다는 생각도 듭니다.

### 궁금한 점
1. Application에 도메인들의 노출
원래 저는 Application에서는 InputView에서 입력한 데이터, OutputView에 보내줄 데이터 (보통 Dto) 만 노출시켰었습니다. 이번에 Dto를 삭제해주면서 도메인이 많이 선언이 되었는데 괜찮을까요??

Hongdonggeon committed:
refactor : 수익률 계산 기능 ResultStatus로 역할 위임

Hongdonggeon committed:
remove : 불필요한 메서드 삭제

lxxjn0 commented on src/main/java/utils/StringUtils.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class StringUtils {
```
해당 클래스의 메서드들이 모두 한 곳에서 한 번씩만 사용되는 것으로 보이는데, 이런 경우에는 굳이 클래스로 관리하기 보다는 사용되는 곳에서 메서드로 관리되는 것이 더 좋다고 생각합니다. 

유틸성 클래스는 말 그대로 도구 클래스, 즉 많은 곳에서 비즈니스 로직을 가지지 않는 동일한 기능으로 사용되기를 기대할 때 사용하면 적절하다고 생각하는데 지금처럼 한 곳에서 한 번만 사용한다면 따로 분리하지 않아도 될 것 같고 이후 추가적인 사용 용도가 생겼을 시점에 분리하면 좋을 것 같아요!! 👍🏼

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumber.java:
```diff
@@ -37,17 +36,26 @@ public static LottoNumber valueOf(int number) {
         return number;
     }
+
+    @Override
+    public String toString() {
+        return "" + number;
+    }
```
toString의 사용 용도에 대해서 한번 학습해보면 좋을 것 같아요!! 👍🏼

lxxjn0 commented on src/main/java/domain/Lotto/LottoFactory.java:
```diff
@@ -0,0 +1,57 @@
+        Collections.shuffle(lottoNumbers);
+
+        List<LottoNumber> completedLottoNumbers = lottoNumbers.stream()
+                .limit(LOTTO_SIZE)
+                .sorted(Comparator.comparing(LottoNumber::getNumber))
+                .collect(Collectors.toList());
+
+        return new Lotto(completedLottoNumbers);
+    }
+
+    public static Lotto generateManualLotto(List<Integer> numbers) {
+        List<LottoNumber> lottoNumbers = numbers.stream()
+                .sorted()
+                .map(LottoNumber::valueOf)
+                .collect(Collectors.toList());
+
+        return new Lotto(lottoNumbers);
+    }
+
+    public static Lotto generateManualLotto(String numbers) {
+        List<String> numberValues = StringUtils.toStrings(numbers);
+        List<Integer> lottoNumbers = StringUtils.toNumbers(numberValues);
+        return generateManualLotto(lottoNumbers);
+    }
+
+    public static List<Lotto> generateAutoLottos(int autoLottoCount) {
+        List<Lotto> autoLottos = new ArrayList<>();
+        for (int i = 0; i < autoLottoCount; i++) {
+            autoLottos.add(generateAutoLotto());
+        }
+        return autoLottos;
+    }
+
+    public static List<Lotto> generateManualLottos(List<String> manualLottoValues) {
+        List<Lotto> manualLottos = new ArrayList<>();
+        for (String manualLottoValue : manualLottoValues) {
+            manualLottos.add(generateManualLotto(manualLottoValue));
+        }
+        return manualLottos;
+    }
```
위에서 아래로 신문을 읽듯이 메서드를 배치를 한다면 가독성을 높일 수 있습니다!! (이는 곧 사용하는 메서드와 사용되는 메서드의 거리가 가까운 것이 좋다는 말과 동일합니다 😄 )

lxxjn0 commented on src/main/java/Application.java:
```diff
@@ -1,22 +1,26 @@
         InputView inputView = InputView.getInstance();
         OutputView outputView = OutputView.getInstance();
         LottoController lottoController = new LottoController();
 
-        LottosDto lottosDto = lottoController.purchase(inputView.inputPurchaseAmount());
-        outputView.printPurchasedLotto(lottosDto);
-
-        lottoController.determineWinningNumber(inputView.inputWinningNumber(), inputView.inputBonusBall());
+        Money money = lottoController.chargeMoney(inputView.inputPurchaseAmount());
+        LottoCount lottoCount = lottoController.selectLottoCount(money, inputView.inputManualLottoCount());
+        Lottos lottos = lottoController.purchaseLotto(lottoCount, inputView.inputManualLottoNumber(lottoCount.getManualLottoCount()));
+        outputView.printPurchasedLotto(lottoCount.getManualLottoCount(), lottoCount.getAutoLottoCount(), lottos);
 
-        outputView.printResult(lottoController.makeResult());
+        WinningLotto winningLotto = lottoController.determineWinningNumber(inputView.inputWinningNumber(), inputView.inputBonusBall());
+        ResultStatus resultStatus = lottoController.makeResult(lottos, winningLotto);
+        outputView.printResult(resultStatus.getResultStatistics());
+        outputView.printIncomeRate(lottoController.calculateImcomeRate(resultStatus, money));
     }
```
Controller에서 비즈니스 로직을 관리하고 Application이 외부에서의 입력을 controller로 전달한다면 어쩔 수 없이 도메인이 Application에 노출되게 된다고 생각합니다. DTO를 없앴기 때문이죠

그래서 현재는 Application에 도메인이 노출되더라도 큰 문제는 없을 것 같아요!! (이후 웹 애플리케이션에서는 dto를 사용하면서 자연스럽게 해결될 문제라고 생각합니다 👍🏼 )

lxxjn0 reviewed (approved):
쿤, 리뷰 반영하느라 수고 많으셨습니다 👍🏼 

이번 미션이 쉬운 미션은 아니었을 텐데 구현하고 피드백 반영하느라 고생 많으셨어요!! 처음 미션을 진행했을 때에 비해서 훨씬 더 코드가 객체지향적이고 책임과 행위가 적절히 분리된 것 같아요!! 👍🏼 

이번 로또 미션 진행하느라 수고 많으셨고 해당 미션은 여기서 머지하도록 하겠습니다 😄 리뷰 반영하느라 정말 수고 많으셨고 다음에 일취월장한 쿤을 다시 만날 수 있길 기대할게요 :)
